### DISCUSSION TITLE
How to display progress on IterableDataset?

### DISCUSSION THREAD

**Discussion by rongcuid** [Lightning-AI/pytorch-lightning#7528] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
When I use an `IterableDataset`, the `Trainer.fit()` gives something like:

    Epoch 0: : 21it [02:41,  7.69s/it, loss=0.663, v_num=18]  

Which does not display a progress. Even if I specify a validation interval, there is no progress shown. Is there a way I can specify an "epoch length", or something like that for `IterableDataset`?

**Comment by awaelchli** [comment, marked as answer]
Hi
IterableDatasets don't have a lengh defined. They can either be infinite in size or raise StopIteration at an arbitrary point. Hence, there is no notion of epochs and we cannot anticipate when it finishes iterations. 

If you really need to have epochs, I suggest you try to rewrite your dataset logic into a regular Dataset.

**Reply by rongcuid** [reply]
Hi,
I am in fact using WebDataset, which itself is an IterableDatset. This is done for performance reasons, when WDS runs twice faster than the cached Daaset and >100 times faster than the uncached Dataset. Therefore, I am unlikely to change to a regular dataset.

Is there a way to somehow specify an "epoch" size to be displayed, but still use the whole dataset? For example, i know my dataset has about 2500 batches. Can I just specify 2500 as the epoch size?

**Reply by awaelchli** [reply]
You can subclass your dataset and define the len, if you can determine it. 


```python
import os

import torch
from torch.utils.data import DataLoader, IterableDataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(IterableDataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    # -----------
    # DEFINE THIS
    # -----------
    def __len__(self):
        return self.len

    def __iter__(self):
        for x in self.data:
            yield x


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        return {"loss": loss}

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64000), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        num_sanity_val_steps=0,
        weights_summary=None,
    )
    trainer.fit(model, train_dataloader=train_data)


if __name__ == '__main__':
    run()

```

**Reply by awaelchli** [reply]
`Epoch 0:  28%|██▊       | 8878/32000 [00:07<00:19, 1169.46it/s, loss=-7.08e+03, v_num=21]`

**Reply by rongcuid** [reply]
That sounds nice. However, if I subclass it, do I need to do special things to prevent the dataset from returning the same items when I use `num_workers > 0` in `DataLoader`? I have not looked into `WebDataset`'s code yet so I don't know how it handles this.

**Reply by awaelchli** [reply]
In webdataset, apparently you can set the length:
https://github.com/tmbdev/webdataset/blob/3431eeada3cba81c19966604116c4b5d6d6994ae/webdataset/dataset.py#L272
It's not documented.

**Reply by rongcuid** [reply]
Oh, that is great! Thank you for your help.

**Reply by rongcuid** [reply]
Just so you know, if I define a `__len__` that returns a different value from actual, `DataLoader` will flood the screen with `UserWarning`s.
